Send user updated websocket event on plugin.UpdateUser

chetanyakan opened the pull request:
#### Summary

- Sends a WebSocket event when `plugin.UpdateUser` is called so it is updated immediately for all the users.

#### Release Note

```release-note
- UserUpdate WebSocket Event is now broadcast by 2 more APIs - plugin.UpdateUser and ConvertBotToUser
- Change in a User's NotifyProps during the user import process will not send an email notification. This is done to make it consistent with other parts of the import process where change in user's attributes would also not send any notifications.
```

mattermod commented:
Hello @chetanyakan,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

agnivade commented on app/plugin_api.go:
```diff
@@ -279,7 +279,14 @@ func (api *PluginAPI) DeletePreferencesForUser(userId string, preferences []mode
+		return nil, err
+	}
+
+	api.app.sendUpdatedUserEvent(*updatedUser)
```
We should move this inside the `UpdateUser` method itself. There are 2 places where `sendUpdatedUserEvent` is called after UpdateUser. We should remove those as well.

chetanyakan replied:
@agnivade This will send the UpdatedUser WebSocket event in other places a well like - 
`[UpdateUserNotifyProps](https://github.com/mattermost/mattermost-server/blob/c1dd23a3c80365c4ca4e2721e55c0c2c90038b25/app/user.go#L1348)` and `[ConvertBotToUser](https://github.com/mattermost/mattermost-server/blob/c1dd23a3c80365c4ca4e2721e55c0c2c90038b25/app/user.go#L2334)`

Would that be an acceptable change?

agnivade replied:
Yes, I think that should be fine. Thanks.

chetanyakan replied:
@agnivade done

agnivade reviewed (changes requested):
Thanks @chetanyakan. Please take a look at the comments.

agnivade commented:
/update-branch

mattermod commented:
We don't have permissions to update this PR, please contact the submitter to apply the update.

agnivade commented:
Woops, @chetanyakan - please update the PR with the latest master. Thanks!

agnivade commented:
Also, let's flesh out the release notes properly by specifically mentioning which 2 API requests will now broadcast the user update messages via websocket.

chetanyakan commented:
Done @agnivade

agarciamontoro reviewed (approved):
Thank you for your contribution, @chetanyakan :tada:

(and sorry for the late review, I was out on vacation!)

lindy65 commented:
Hi @agnivade / @agarciamontoro - is there a Jira ticket for this PR? Please link me to it? Thanks!

agarciamontoro commented:
Hm, that's a good question I don't an answer for, I had not noticed when I reviewed it. Maybe @agnivade or @chetanyakan know.

agnivade commented:
There is no JIRA ticket @lindy65. This is something @chetanyakan created on his own. I am not sure of the QA process we follow for contributor generated PRs.

lindy65 commented:
Hi @agnivade - would you help create a Jira ticket please?  This is required so that once the PR is merged, we don't lose track of it and have something to test against from resolved tickets.  Please also help add test steps to the ticket? TIA :)

agnivade reviewed (changes requested):
Actually, @chetanyakan - apologies for asking you to do this again, but on a closer look, I realized that we don't want to do this for UpdateNotifyProps. That's because it happens during the user import process and there is no need to send any notifications.

It should not be too much work. The method is just used in 2 places. We need to add another flag to `UpdateUserNotifyProps` for `sendNotifications` and pass false from app/import_functions.go.

And only send the user event if `sendNotifications` is set to true.

Everything else remains same. Sorry for the double work here.

agnivade commented:
@lindy65 - https://mattermost.atlassian.net/browse/MM-31915 is the ticket, but please hold off from testing for now. The PR needs to be updated a bit.

lindy65 commented:
Thanks @agnivade - I'll hold off testing until you give me the go-ahead.  Can you tell me what plugin updates a user profile so I can test using it?

agnivade commented:
There are many like that but probably @agarciamontoro knows a suitable one to test this.

agarciamontoro commented:
I cannot think of one right now, actually :thinking: I checked that the Github plugin does call UpdateUser when disconnecting the Github account: https://github.com/mattermost/mattermost-plugin-github/blob/8f2addacdcf24636d697fc2dc015cbadbcbb2c8a/server/plugin/plugin.go#L331, but I'm not sure if that's very userful for your testing purposes, @lindy65.

lindy65 commented:
Thanks @agarciamontoro :)  Perhaps I should test this after merge where it'll be easier with available plugins on `master`

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

agnivade commented:
Hey @chetanyakan - how is it going with this one? Let us know if you need any help.

chetanyakan commented:
Sorry, @agnivade, I have been really busy. Will update this shortly.

chetanyakan commented:
> Actually, @chetanyakan - apologies for asking you to do this again, but on a closer look, I realized that we don't want to do this for UpdateNotifyProps. That's because it happens during the user import process and there is no need to send any notifications.
> 
> It should not be too much work. The method is just used in 2 places. We need to add another flag to `UpdateUserNotifyProps` for `sendNotifications` and pass false from app/import_functions.go.
> 
> And only send the user event if `sendNotifications` is set to true.
> 
> Everything else remains same. Sorry for the double work here.

Hey @agnivade just to clarify, I will be making the following changes:

1. Update the `app.UpdateUser` method to call `a.sendUpdatedUserEvent(*user)` only when the value of the second parameter `sendNotifications` is true.
2. Update the method `app.UpdateUserNotifyProps` to add an extra parameter `sendNotifications` and pass that to `UpdateUser`.
3. Update the usages of `UpdateUserNotifyProps` method to pass the values of  `sendNotifications` and pass `false` in `app/import_functions.go`.

Is that correct?

agnivade commented:
That is correct! A small note there is that `app/import_functions.go` is the only place that calls `UpdateUserNotifyProps`.

I should add here for posterity that this change will make `UpdateUserNotifyProps` not send mail notifications from import_functions.go. But there are other areas, where `UpdateUser` / `UpdateUserRoles` are called during the import process with notifications disabled. And it sounds like the correct process to me. So we would be actually making the behavior more consistent.

And for full transparency, we should add this in release notes too, that change in a user's notify props during the import process will not send a mail notification. This is done to make it consistent with changes in other attributes of the user.

agnivade commented:
Are the release notes still correct? `UpdateUserNotifyProps` won't broadcast an event any more correct?

/update-branch

mattermod commented:
We don't have permissions to update this PR, please contact the submitter to apply the update.

agnivade commented:
Woops, please update the PR with the latest master.

chetanyakan commented:
@agnivade I have updated the PR to the head of `master` and updated the release notes.

Your note was correct, `UpdateUserNotifyProps` won't broadcast the user updated event.

agnivade reviewed (approved):
Nice work @chetanyakan !

agnivade commented:
@lindy65 - over to you :)

lindy65 commented:
> I cannot think of one right now, actually 🤔 I checked that the Github plugin does call UpdateUser when disconnecting the Github account: https://github.com/mattermost/mattermost-plugin-github/blob/8f2addacdcf24636d697fc2dc015cbadbcbb2c8a/server/plugin/plugin.go#L331, but I'm not sure if that's very userful for your testing purposes, @lindy65.

Hey @agnivade - would this PR be better tested after merge based on what `@agarciamontoro` mentioned above?  I'm not sure which plugin tests this fix properly but there is more choice on `community-daily`? Thoughts?

agnivade commented:
I agree.

lindy65 reviewed (approved):
This will be tested on `community-daily` after merge 👍
